File: app/src/main/java/com/mouredev/weeklychallenge2022/Challenge35.kt
package com.mouredev.weeklychallenge2022

/*
 * Reto #35
 * BATALLA POKÉMON
 * Fecha publicación enunciado: 29/08/22
 * Fecha publicación resolución: 06/09/22
 * Dificultad: MEDIA
 *
 * Enunciado: Crea un programa que calcule el daño de un ataque durante una batalla Pokémon.
 * - La fórmula será la siguiente: daño = 50 * (ataque / defensa) * efectividad
 * - Efectividad: x2 (súper efectivo), x1 (neutral), x0.5 (no es muy efectivo)
 * - Sólo hay 4 tipos de Pokémon: Agua, Fuego, Planta y Eléctrico (buscar su efectividad)
 * - El programa recibe los siguientes parámetros:
 *  - Tipo del Pokémon atacante.
 *  - Tipo del Pokémon defensor.
 *  - Ataque: Entre 1 y 100.
 *  - Defensa: Entre 1 y 100.
 *
 * Información adicional:
 * - Usa el canal de nuestro Discord (https://mouredev.com/discord) "🔁reto-semanal"
 *   para preguntas, dudas o prestar ayuda a la comunidad.
 * - Tienes toda la información sobre los retos semanales en
 *   https://retosdeprogramacion.com/semanales2022.
 *
 */

fun main(){
    val odish = Pokemon(name = "Odish", elementType = Pokemon.Element.Grass,
        attackPower = 100, defencePower = 30, movePower = 50)

    val lechonk = Pokemon(name = "Lechonk", elementType = Pokemon.Element.Fire,
        attackPower = 50, defencePower = 80, movePower = 90)

    attack(odish, lechonk)
    attack(lechonk, odish)
}

class Pokemon(val name : String, val elementType : Element,
               val attackPower : Int, val defencePower : Int, val movePower : Int){

    sealed class Element(
        val weak : List<Element>,
        val effective : List<Element>,
        val resist : List<Element>,
        val immune : List<Element>? = null){
        object Electric : Element(
            weak = listOf(Ground),
            effective = listOf(Water, Flying),
            resist = listOf(Electric, Flying, Steel))

        object Water : Element(
            weak = listOf(Grass, Electric),
            effective = listOf(Ground, Rock, Fire),
            resist = listOf(Fire, Ice, Steel, Water))

        object Fire : Element(
            weak = listOf(Water, Ground, Rock),
            effective = listOf(Grass, Steel, Ice, Bug),
            resist = listOf(Steel, Fire, Grass, Ice, Bug))

        object Grass : Element(
            weak = listOf(Fire, Ice, Flying, Poison, Bug),
            effective = listOf(Water, Ground, Rock),
            resist = listOf(Water, Grass, Electric, Ground))

        object Ground : Element(
            weak = listOf(Water, Ice, Grass),
            effective = listOf(Fire, Electric, Poison, Rock, Steel),
            resist = listOf(Rock, Poison),
            immune = listOf(Electric))

        object Ice : Element(
            weak = listOf(Fire, Fighting, Steel, Rock),
            effective = listOf(Grass, Flying, Dragon, Ground),
            resist = listOf(Ice))

        object Steel : Element(
            weak = listOf(Ground, Fire, Fighting),
            effective = listOf(Ice, Rock),
            resist = listOf(Psychic, Dragon, Steel, Bug,
                Ice, Normal, Grass, Flying, Rock))

        object Rock : Element(
            weak = listOf(Ground, Water, Grass, Steel, Fighting),
            effective = listOf(Ice, Fire, Bug, Flying),
            resist = listOf())

        object Poison : Element(
            weak = listOf(Ground, Psychic),
            effective = listOf(Grass),
            resist = listOf(Poison, Bug, Fighting, Grass))

        object Flying : Element(
            weak = listOf(Rock, Ice, Electric),
            effective = listOf(Grass, Fighting, Bug),
            resist = listOf(Grass, Fighting, Bug),
            immune = listOf(Ground))

        object Dark : Element(
            weak = listOf(Fighting, Bug),
            effective = listOf(Psychic, Ghost),
            resist = listOf(Psychic, Dark, Ghost))

        object Normal : Element(
            weak = listOf(Fighting),
            effective = listOf(),
            resist = listOf(),
            immune = listOf(Ghost))

        object Ghost : Element(
            weak = listOf(Ghost, Dark),
            effective = listOf(Psychic, Ghost),
            resist = listOf(Poison, Bug),
            immune = listOf(Normal, Fighting))

        object Fighting : Element(
            weak = listOf(Flying, Psychic),
            effective = listOf(Ice, Rock, Steel, Normal, Dark),
            resist = listOf(Dark, Rock, Bug))

        object Dragon : Element(
            weak = listOf(Dragon, Ice),
            effective = listOf(Dragon),
            resist = listOf(Fire, Water, Electric, Grass))

        object Bug : Element(
            weak = listOf(Rock, Fire, Flying),
            effective = listOf(Psychic, Grass, Dark),
            resist = listOf(Fighting, Grass, Ground))

        object Psychic : Element(
            weak = listOf(Bug, Dark, Ghost),
            effective = listOf(Fighting, Poison),
            resist = listOf(Psychic, Fighting))

        fun getEffectiveMultiplier(attackElement : Element) : Float {
            return if (immune != null && immune.contains(attackElement))
                0.0f
            else if (resist.contains(attackElement))
                0.5f
            else if (weak.contains(attackElement))
                2.0f
            else
                1.0f
        }
    }
}

fun attack(pokemonA : Pokemon, pokemonD : Pokemon){

    val damage = pokemonA.movePower*(pokemonA.attackPower/pokemonD.defencePower.toFloat())
    val effectiveMultiplier = pokemonD.elementType.getEffectiveMultiplier(pokemonA.elementType)

    println("\n ${pokemonA.name} attacks ${pokemonD.name}.")
    println("It did ${damage*effectiveMultiplier} points of damage.")
    println(when(effectiveMultiplier){
        0.0f -> "${pokemonD.name} is immune."
        0.5f -> "It was not effective"
        2.0f -> "It was super effective"
        else -> ""
    })
}

File: app/src/main/java/com/mouredev/weeklychallenge2022/Challenge35.cpp
 /*package com.mouredev.weeklychallenge2022


#  * Reto #35
#  * BATALLA POKÉMON
#  * Fecha publicación enunciado: 29/08/22
#  * Fecha publicación resolución: 06/09/22
#  * Dificultad: MEDIA
#  *
#  * Enunciado: Crea un programa que calcule el daño de un ataque durante una batalla Pokémon.
#  * - La fórmula será la siguiente: daño = 50 * (ataque / defensa) * efectividad
#  * - Efectividad: x2 (súper efectivo), x1 (neutral), x0.5 (no es muy efectivo)
#  * - Sólo hay 4 tipos de Pokémon: Agua, Fuego, Planta y Eléctrico (buscar su efectividad)
#  * - El programa recibe los siguientes parámetros:
#  *  - Tipo del Pokémon atacante.
#  *  - Tipo del Pokémon defensor.
#  *  - Ataque: Entre 1 y 100.
#  *  - Defensa: Entre 1 y 100.
#  *
#  * Información adicional:
#  * - Usa el canal de nuestro Discord (https://mouredev.com/discord) "??reto-semanal"
#  *   para preguntas, dudas o prestar ayuda a la comunidad.
#  * - Tienes toda la información sobre los retos semanales en
#  *   https://retosdeprogramacion.com/semanales2022.
#  *
#  */

#include <iostream>
using namespace std;


float atac(string atacante,string defensor){

	float efectividad;
//fuego	
	if((atacante=="fuego" && defensor=="agua")||(atacante==defensor)){
		
		efectividad=0.5;
	}
	if(atacante=="fuego" && defensor=="planta"){
		efectividad=2;
	}
	if(atacante=="fuego" && defensor=="electrico"){
		efectividad=1;
	}
	
//agua
	if((atacante=="agua"&&defensor=="planta")||(atacante==defensor)){
		efectividad=0.5;
	}
	if(atacante=="agua" && defensor=="fuego"){
		efectividad=2;
	}
	if(atacante=="agua" && defensor=="electrico"){
		efectividad=1;
	}	
//electrico
	if((atacante=="electrico"&&defensor=="planta")||(atacante==defensor)){
		efectividad=0.5;
	}
	if(atacante=="electrico" && defensor=="agua"){
		efectividad=2;
	}
	if(atacante=="electrico" && defensor=="fuego"){
		efectividad=1;
	}	
//planta
	if((atacante=="planta"&&defensor=="fuego")||(atacante==defensor)){
		efectividad=0.5;
	}
	if(atacante=="planta" && defensor=="agua"){
		efectividad=2;
	}
	if(atacante=="planta" && defensor=="electrico"){
		efectividad=1;
	}
	return efectividad;
}


int main(int argc, char** argv) {
	
	string atacante="";
	string defensor="";
	float ataque;
	float defensa;
	int tipo1;
	int tipo2;
	double attack;
	float efectividad;
	
	cout<<"Seleccione el tipo del pokemon atacante: "<<endl;
	cout<<"1. Fuego "<<endl;
	cout<<"2. Agua "<<endl;
	cout<<"3. Planta "<<endl;
	cout<<"4. Electrico "<<endl;	
	cin>>tipo1;
	
	switch (tipo1){
		case 1: atacante="fuego";
		break;
		case 2: atacante="agua";
		break;
		case 3: atacante="planta";
		break;
		case 4: atacante="electrico";
		break;
		default:
			break;
	}
		
		
	system("cls");
	
	cout<<"Seleccione el tipo del pokemon defensor: "<<endl;
	cout<<"1. Fuego "<<endl;
	cout<<"2. Agua "<<endl;
	cout<<"3. Planta "<<endl;
	cout<<"4. Electrico "<<endl;	
	cin>>tipo2;
	
	switch (tipo2){
		case 1: defensor="fuego";
		break;
		case 2: defensor="agua";
		break;
		case 3: defensor="planta";
		break;
		case 4:	defensor="electrico";
		break;
		default:
			break;
	}	
	
	efectividad=atac(atacante,defensor);
	
		system("cls");

	cout<<"Seleccione la cantidad de ataque Entre 1 y 100: "<<endl;
	cin>>ataque;
	cout<<"Seleccione la cantidad de defensa Entre 1 y 100: "<<endl;
	cin>>defensa;
	
	
	
	attack = 50*(ataque/defensa)*efectividad;
	
	cout<<"Daño de ataque: "<<attack<<endl;
	
	return 0;
}
